packages/eslint-plugin/src/rules/no-floating-promises.ts

```diff
@@ -296,12 +327,31 @@ export default createRule<Options, MessageId>({
   },
 });
 
+function isPromiseArray(checker: ts.TypeChecker, node: ts.Node): boolean {
+  const type = checker.getTypeAtLocation(node);
+  for (const ty of tsutils
+    .unionTypeParts(type)
+    .map(t => checker.getApparentType(t))) {
+    if (checker.isArrayType(ty)) {
+      const arrayType = checker.getTypeArguments(ty)[0];
```

Comment:
This doesn't appear to be a sound way to get the element type. A few lines in the test cases have more than one type argument, but I don't know why. For example, 
```
function f<T extends Array<Promise<number>>>(a: T): void {
  a;
}
```

Reply:
Yeah it's a little spooky but you should be good. If `ty` is an array type then the first type argument should be matched to the array type parameter. Remaining type arguments get assigned to whatever the element type is (here, `Promise<...>`).

Reply:
Hmm.. I guess, I don't totally follow, since
```
checker.getTypeArguments(/* node corresponding to a */) // results in [Promise, T],
```
whereas I think I would expect `[T, Promise]`, or `[Promise, number]` depending on how I interpret the wording of your comment.

However I'm perfectly happy to take your word for it if you think it's good :)

Reply:
I have never been more grateful for test infrastructure :)